### PR TITLE
Add `--limit` flag to `pulumi stack history` and consume from automation api

### PR DESCRIPTION
**Pull request by EvanBoyle**
We had some reports of [perf issues](https://github.com/pulumi/pulumi/issues/5453#issuecomment-773442294) related to pulumi stack history. Specifically we use this command to fetch update summary after an update with automation API. As you perform more updates, performance degrades as we fetch more data and decrypt more secrets. This change does two things:

1. add a `--limit=N` flag to `pulumi history` and `pulumi stack history` to allow only fetching the most recent `N` records.
2. consume this flag in the appropriate places in the Automation API SDKs for node, go, and python. 

Note that this is technically a small breaking change for Go (other languages mark this as optional and previous SDK shape is still valid) but this is within the boundaries we've set for the alpha and is an important change to make for performance. 

We don't have any infrastructure setup to test the filestate backend changes. There are tests that create stacks, select stacks, but none that go through the lifecycle of create, do an update, etc. Adding this was outside of what I have bandwidth for, but I tested this workflow locally with the filestate backend. See https://github.com/pulumi/pulumi/issues/3760. 

Fixes https://github.com/pulumi/pulumi/issues/6253
Fixes https://github.com/pulumi/pulumi/issues/6252

Note this will need to be ported to C# after that lands.

### CODE REVIEW

**Review by komalali — APPROVED** [review]
LGTM except for a funky conditional.

File: pkg/backend/filestate/state.go [hunk 1]
```diff
@@ -327,9 +331,16 @@ func (b *localBackend) getHistory(name tokens.QName) ([]backend.UpdateInfo, erro
 
 	var updates []backend.UpdateInfo
 
+	if limit > 0 {
+		if limit > len(allFiles) {
+			limit = len(allFiles)
+		}
+		limit = len(allFiles) - limit
```
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply pkg/backend/filestate/state.go:338]: The condition is a little funky here, but it is written as intended. The first condition is the equivalent of `min(limit, len(allFiles)` incase the limit is larger than the total number of results:

```go
		if limit > len(allFiles) {
			limit = len(allFiles)
		}
```

After that, we adjust the loop condition:

```go
limit = len(allFiles) - limit
```

[the remark this answers is not in the document]
Comment (komalali) [reply pkg/backend/filestate/state.go:338]:
```
limit = 10, len(allFiles) = 5

if 10 > 5 {
	limit = 5
}
limit = 5 - 5 = 0
```

So if the limit is larger than the total number of results we would end up with `limit=0`?

[the remark this answers is not in the document]
Comment (komalali) [reply pkg/backend/filestate/state.go:338]: Ah, I get it after looking at the code that follows.

**Comment by cnunciato** [comment]
Does this also (kinda) fix https://github.com/pulumi/pulumi/issues/6168?

**Comment by komalali** [comment]
@cnunciato it sure (kinda) does, thanks for bringing it up!

**Pull request merged**